### PR TITLE
JDK-8247795 allow factory methods for inline types to return j.l.Obje…

### PR DESCRIPTION
**Pull request by hseigel**
Expect the return type for the <init> factory methods of hidden inline classes to return type java.lang.Object.  Also, fix issue so that Reflection::new_constructor(), not new_method(), is called for <init> factory methods.

---------
### Progress
- [x] Change must not contain extraneous whitespace

### Issue
 * [JDK-8247795](https://bugs.openjdk.java.net/browse/JDK-8247795): [lworld] Fix VM support of factory methods for inline hidden and UAC classes ⚠️ Title mismatch between PR and JBS.

### Reviewers
 * Frederic Parain ([fparain](@fparain) - Committer)

### Download
`$ git fetch https://git.openjdk.java.net/valhalla pull/94/head:pull/94`
`$ git checkout pull/94`

### CODE REVIEW

**Comment by mlchung** [comment]
I consider this fix is a workaround to allow prototyping inline hidden class (for example to convert lambda proxy class to inline class).   We need to sort out JVM spec issues - what return types should be allowed?  JDK-8224057 is a companion issue.

**Comment by mlchung** [comment]
@hseigel - have you seen my feedback?  You integrated it as is?

**Comment by mlchung** [comment]
> @hseigel - have you seen my feedback? You integrated it as is?

Sigh... I just realized I didn't click "submit review".    You can follow up my comments in a separate patch.  Sorry for not posting my review comments properly.

File: src/hotspot/share/classfile/classFileParser.cpp [hunk 1]
```diff
@@ -2479,9 +2479,9 @@ Method* ClassFileParser::parse_method(const ClassFileStream* const cfs,
       if (ss.is_reference()) {
         Symbol* ret = ss.as_symbol();
         const Symbol* required = class_name();
-        if (is_unsafe_anonymous()) {
-          // The original class name in the UAC byte stream gets changed.  So
-          // using the original name in the return type is no longer valid.
+        if (is_hidden() || is_unsafe_anonymous()) {
```
[the remark this answers is not in the document]
Comment (hseigel) [reply src/hotspot/share/classfile/classFileParser.cpp:2482]: VM anonymous classes need to be supported until they are obsoleted.  Removing this call to is_unsafe_anonymous() can easily be done once that happens.

File: src/hotspot/share/runtime/reflection.cpp [hunk 2]
```diff
@@ -1222,7 +1222,13 @@ oop Reflection::invoke_constructor(oop constructor_mirror, objArrayHandle args,
   if (!method->signature()->is_void_method_signature()) {
     assert(klass->is_inline_klass(), "inline classes must use factory methods");
     Handle no_receiver; // null instead of receiver
-    return invoke(klass, method, no_receiver, override, ptypes, T_VALUETYPE, args, false, CHECK_NULL);
+    BasicType rtype;
+    if (klass->is_hidden() || klass->is_unsafe_anonymous()) {
```
[the remark this answers is not in the document]
Comment (hseigel) [reply src/hotspot/share/runtime/reflection.cpp:1226]: See above comment about is_unsafe_anonymous().  I can add a comment about the JVM Spec issue in a follow-on fix.

File: test/hotspot/jtreg/runtime/valhalla/valuetypes/HiddenInlineClassTest.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+import java.nio.file.Path;
+import java.nio.file.Paths;
+
+public class HiddenInlineClassTest {
+
+    static final Path CLASSES_DIR = Paths.get(System.getProperty("test.classes"));
+    String hello = "hello";
+
+    static byte[] readClassFile(String classFileName) throws Exception {
+        File classFile = new File(CLASSES_DIR + File.separator + classFileName);
```
[the remark this answers is not in the document]
Comment (hseigel) [reply test/hotspot/jtreg/runtime/valhalla/valuetypes/HiddenInlineClassTest.java:48]: Multiple tests could benefit by using the NIO change that you suggest.  I'll enter an RFE for that.

File: test/hotspot/jtreg/runtime/valhalla/valuetypes/HiddenInlineClassTest.java [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+        }
+    }
+
+    public static void main(String[] args) throws Throwable {
+        Lookup lookup = MethodHandles.lookup();
+        byte[] bytes = readClassFile("HiddenPoint.class");
+
+        // Define a hidden class that is an inline type.
+        Class<?> c = lookup.defineHiddenClass(bytes, true, NESTMATE).lookupClass();
+        Object hp = c.newInstance();
```
[the remark this answers is not in the document]
Comment (hseigel) [reply test/hotspot/jtreg/runtime/valhalla/valuetypes/HiddenInlineClassTest.java:65]: Is it a problem that it's a nestmate?

[the remark this answers is not in the document]
Comment (mlchung) [reply test/hotspot/jtreg/runtime/valhalla/valuetypes/HiddenInlineClassTest.java:65]: This is a copy-n-paste issue.  I expect a test knows what it does and it should use it with a clear intention.  I prefer to take it out or add a comment explaining why NESTMATE is used for this hidden inline test.

File: src/hotspot/share/classfile/classFileParser.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (mlchung) [reply src/hotspot/share/classfile/classFileParser.cpp:2482]: I do not agree that we need to make VM anonymous classes to support inline types.    This block is for inline types only.  Please remove it.

File: src/hotspot/share/runtime/reflection.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (mlchung) [reply src/hotspot/share/runtime/reflection.cpp:1226]: Please do not make VM anonymous classes to support inline type.

**Pull request closed without merging**